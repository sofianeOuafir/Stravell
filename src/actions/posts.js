import database from "../firebase/firebase";

const addPost = post => ({
  type: "ADD_POST",
  post
});

const startAddPost = (postData = {}) => {
  return (dispatch, getState) => {
    const { uid, userName, userPhotoURL } = getState().auth;
    const { title, body, description, image, createdAt, updatedAt } = postData;
    const post = {
      uid,
      userName,
      userPhotoURL,
      title,
      description,
      image,
      body,
      createdAt,
      updatedAt
    };
    return database
      .ref("posts")
      .push(post)
      .then(ref => {
        dispatch(
          addPost({
            id: ref.key,
            ...post
          })
        );
      })
      .catch(e => console.log(e));
  };
};

const editPost = ({ id, updates }) => ({
  type: "EDIT_POST",
  id,
  updates
});

const startEditPost = ({ id, updates }) => {
  return dispatch => {
    database
      .ref(`posts/${id}`)
      .update(updates)
      .then(ref => {
        dispatch(editPost({ id, updates }));
      });
  };
};

const setPosts = posts => ({
  type: "SET_POSTS",
  posts
});

const startSetPosts = () => {
  return dispatch => {
    return new Promise((resolve, reject) => {
      let posts = [];
      database
      .ref("posts")
      .orderByChild("createdAt")
      .on("value", snapshot => {
        snapshot.forEach(snapshotChild => {
          posts.push({
            id: snapshotChild.key,
            ...snapshotChild.val()
          });
        });
        posts = posts.reverse();
        dispatch(setPosts(posts));
        resolve();
      });
      
    })
  };
};

export { startEditPost, startAddPost, startSetPosts };
