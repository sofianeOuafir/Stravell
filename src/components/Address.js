import React from "react";
import { MdMyLocation } from "react-icons/md";
import Link from "next/link";
import { slugify } from "underscore.string";

import { getPlaceIdFromLatLng } from "./../lib/utils/place";

const Address = ({ iconClassName, addressClassName, address, lat, lng }) => {
  const placeId = getPlaceIdFromLatLng({ lat, lng });
  return (
    <div className="flex align-items--center">
      <Link
        href={`/place?id=${placeId}`}
        as={`/place/${slugify(address)}/${placeId}`}
      >
        <div>
          <MdMyLocation className={iconClassName ? iconClassName : ""} />
          <span className={addressClassName ? addressClassName : ""}>
            {address}
          </span>
        </div>
      </Link>
    </div>
  );
};

export default Address;
