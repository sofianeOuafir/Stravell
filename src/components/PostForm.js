import React from "react";
import moment from "moment";
import { EditorState, convertFromRaw, convertToRaw } from "draft-js";
import { FaCheck, FaTimes } from "react-icons/fa";
import uuid from 'uuid';
import MyEditor from "./MyEditor";
import { uploadFile } from "./../aws/s3";
import Loading from "./Loading";
import {
  formatTitle,
  formatDescription,
  getTitleError,
  getDescriptionError,
  getImageError,
  getBodyError
} from "./../lib/utils/post";

class PostForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      title: (props.post && props.post.title) || "",
      description: (props.post && props.post.description) || "",
      image: (props.post && props.post.image) || "",
      body:
        (props.post &&
          EditorState.createWithContent(
            convertFromRaw(JSON.parse(props.post.body))
          )) ||
        EditorState.createEmpty(),
      createdAt: (props.post && moment(props.post.createdAt)) || moment(),
      s3FolderName: (props.post && props.post.s3FolderName) || uuid(),
      updatedAt: moment(),
      titleError: "",
      descriptionError: "",
      imageError: "",
      bodyError: "",
      imageUploading: false
    };
  }

  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({
      description: description,
      descriptionError: getDescriptionError(description)
    }));
  };

  onTitleChange = e => {
    const title = e.target.value;
    this.setState(() => ({ title: title, titleError: getTitleError(title) }));
  };

  onImageChange = e => {
    const file = new File([e.target.files[0]], `main`, {type: e.target.files[0].type});
    if(!file) {
      return;
    }
    this.setState(() => ({ imageUploading: true }), () => {
      uploadFile({ file, location: `pictures/${this.state.s3FolderName}` })
        .then(({ Location }) => {
          this.setState(() => ({ image: Location, imageError: getImageError(Location), imageUploading: false }));
        })
        .catch(err => {
          alert(err);
          this.setState(() => ({ imageUploading: false }))
        });
    });
  };

  onBodyChange = editorState => {
    const body = editorState.getCurrentContent().getPlainText();
    this.setState(() => ({ body: editorState, bodyError: getBodyError(body) }));
  };

  onSubmit = e => {
    e.preventDefault();
    let errors = {
      title: getTitleError(this.state.title),
      description: getDescriptionError(this.state.description),
      image: getImageError(this.state.image),
      body: getBodyError(this.state.body.getCurrentContent().getPlainText())
    };
    if (
      errors.title === "" &&
      errors.description === "" &&
      errors.image === "" &&
      errors.body === ""
    ) {
      this.props.onSubmit({
        title: formatTitle(this.state.title),
        description: formatDescription(this.state.description),
        image: this.state.image,
        body: JSON.stringify(convertToRaw(this.state.body.getCurrentContent())),
        createdAt: this.state.createdAt.valueOf(),
        updatedAt: this.state.updatedAt.valueOf(),
        s3FolderName: this.state.s3FolderName
      });
    } else {
      this.setState(() => ({
        titleError: errors.title,
        descriptionError: errors.description,
        imageError: errors.image,
        bodyError: errors.body
      }));
    }
  }

  getValidationIcon = (error) => {
    if(error) {
      return <FaTimes className={`icon c-warm-peach`}/>;
    } else {
      return <FaCheck className={`icon c-limegreen`} />;
    }
  }

  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        <div className="form__input-container">
          { this.getValidationIcon(this.state.titleError) }
          <input
            placeholder="Write a title here"
            className="text-input"
            type="text"
            onChange={this.onTitleChange}
            autoFocus
            value={this.state.title}
          />
          {this.state.titleError && <p className="error">{this.state.titleError}</p>}
        </div>
        <div className="form__input-container">
          { this.getValidationIcon(this.state.descriptionError) }
          <input
            placeholder="Write a small description here"
            className="text-input"
            type="text"
            onChange={this.onDescriptionChange}
            value={this.state.description}
          />
          {this.state.descriptionError && <p className="error">{this.state.descriptionError}</p>}
        </div>
        <div className="form__input-container">
          { this.getValidationIcon(this.state.imageError) }  
          <input type="file" accept="image/*" onChange={this.onImageChange} className="text-input" />
          {this.state.imageError && <p className="error">{this.state.imageError}</p>}
          <div className="quarterwidth flex align-items--center justify-content--center mt2">
            { this.state.imageUploading ? 
              (
                <Loading size="small" />
              )
              :
              (
                this.state.image && (
                  <img src={`${this.state.image}`} alt={`${this.state.image}`} className="fullwidth" />
                )
              )
            }
          </div>

        </div>
        <div className="form__input-container">
          { this.state.bodyError && <p className="error">{this.state.bodyError}</p>}
          <MyEditor
            placeholder="Write your article here"
            editorState={this.state.body}
            onChange={this.onBodyChange}
            s3FolderName={this.state.s3FolderName}
          />
        </div>
        <div>
          <button className="button">Save Post</button>
        </div>
      </form>
    )
  }
}

export default PostForm;
