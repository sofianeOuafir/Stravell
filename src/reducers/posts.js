const postsReducer = (state = [], action) => {
  switch (action.type) {
    case "ADD_POST":
      return [action.post, ...state];
    case "EDIT_POST":
      return [
        ...state.map(post => {
          if (post.id === action.id) {
            return {
              ...post,
              ...action.updates
            };
          } else {
            return post;
          }
        })
      ];
    case "SET_POSTS":
      return [
        ...action.posts.map(post => {
          if (post.comments) {
            let comments = [];
            for (let i in post.comments) {
              comments.push(post.comments[i]);
            }
            post.comments = comments;
          }
          return post;
        })
      ];
    case "EDIT_POSTS_USER_PHOTO_URL":
      return [
        ...state.map(post => {
          if (post.uid == action.uid) {
            return {
              ...post,
              userPhotoURL: action.userPhotoURL
            };
          } else {
            return post;
          }
        })
      ];
    case "ADD_POST_COMMENT":
      return [
        ...state.map(post => {
          if (post.id == action.comment.postId) {
            if (!post.comments) {
              post.comments = [];
            }
            // post.comments = post.comments.push(action.comment);
            post.comments = [...post.comments, action.comment];
            return post;
          } else {
            return post;
          }
        })
      ];
    default:
      return state;
  }
};

export default postsReducer;
