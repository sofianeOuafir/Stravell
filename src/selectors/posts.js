const getVisiblePosts = (posts, { text, countryCode }) => {
  text = text.toLowerCase()
  return posts.filter(post => {
    const matchText =
      post.title.toLowerCase().includes(text) ||
      post.description.toLowerCase().includes(text);
    const matchCountry = countryCode === '' || post.countryCode === countryCode
    return matchText && matchCountry;
  });
};

export { getVisiblePosts };
