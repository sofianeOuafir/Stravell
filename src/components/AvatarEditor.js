import React, { Fragment } from "react";
import Editor from "react-avatar-editor";
import { connect } from "react-redux";
import uuid from "uuid";
import Avatar from "react-avatar";
import { IoMdCamera } from "react-icons/io";
import Modal from "react-modal";

import { uploadFile } from "./../aws/s3";
import { updateUserProfilePicture } from "./../queries/user";
import { updateProfilePicture } from "./../actions/auth";
import Loading from "./Loading";

class AvatarEditor extends React.Component {
  constructor(props) {
    super(props);
    const { userPhotoURL: currentImageUrl, uid } = props.user;
    this.state = {
      currentImageUrl,
      uploadedImage: null,
      uid,
      uploading: false,
      modalIsOpen: false
    };
  }

  closeModal = () => {
    this.setState(() => ({ modalIsOpen: false }));
  };

  onImageSave = () => {
    const { uid } = this.state;
    const { updateProfilePicture } = this.props;
    const canvas = this.editor.getImage();
    this.setState(() => ({ uploading: true, modalIsOpen: false }));

    canvas.toBlob(blob => {
      const file = new File([blob], "profile_picture", {
        type: blob.type
      });
      uploadFile({ file, location: `users/${uid}` })
        .then(data => {
          const { Location: newUserPhotoURL } = data;
          this.setState(() => ({
            uploading: false,
            currentImageUrl: `${newUserPhotoURL}?${new Date().getTime()}`
          }));

          updateUserProfilePicture({ uid, newUserPhotoURL });
          updateProfilePicture(`${newUserPhotoURL}?${new Date().getTime()}`);
        })
        .catch(e => {
          this.setState(() => ({ uploading: false }));
          alert(e);
        });
    });
  };

  onImageChange = e => {
    const uploadedImage = e.target.files[0];
    if (!uploadedImage) {
      return;
    }
    this.setState(() => ({ uploadedImage, modalIsOpen: true }));
  };

  onCancel = () => {
    this.setState(() => ({
      modalIsOpen: false
    }));
  };

  triggerFileDialog = () => {
    document.getElementById("imageInput").click();
  };

  setEditorRef = editor => (this.editor = editor);

  render() {
    const {
      uploadedImage,
      currentImageUrl,
      uploading,
      modalIsOpen
    } = this.state;
    return (
      <div
        className="relative flex align-items--center justify-content--center"
        style={{ height: 100, width: 100 }}
      >
        <Fragment>
          <Avatar
            className="absolute pointer"
            src={currentImageUrl}
            round={true}
            size={100}
            onClick={this.triggerFileDialog}
          />
          {uploading ? (
            <div className="absolute">
              <Loading size={40} />
            </div>
          ) : (
            <Fragment>
              <input
                id="imageInput"
                className="hide"
                type="file"
                onChange={this.onImageChange}
              />

              <IoMdCamera
                className="absolute pointer h2 text-white"
                onClick={this.triggerFileDialog}
              />
            </Fragment>
          )}
        </Fragment>

        <Modal
          isOpen={modalIsOpen}
          onRequestClose={this.closeModal}
          className="modal"
          contentLabel="Example Modal"
        >
          <p className="modal__title h3 favourite-font-weight">Edit Media</p>
          <Editor ref={this.setEditorRef} image={uploadedImage} />
          <div className="flex justify-content--between">
            <button className="button" onClick={this.onImageSave}>Save</button>
            <button className="button button--warm-peach" onClick={this.onCancel}>Cancel</button>
          </div>
        </Modal>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  updateProfilePicture: userPhotoURL =>
    dispatch(updateProfilePicture(userPhotoURL))
});

export default connect(
  null,
  mapDispatchToProps
)(AvatarEditor);
