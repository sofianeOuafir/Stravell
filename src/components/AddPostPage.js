import React from "react";
import PostForm from "./PostForm";
import { connect } from "react-redux";
import { startAddPost } from "../actions/posts";
import PageHeader from './PageHeader';

class AddPostPage extends React.Component {
  onSubmit = post => {
    this.props.startAddPost(post);
    this.props.history.push("/dashboard");
  };

  render() {
    return (
      <div>
        <PageHeader title="Create Post" />
        <div className="content-container">
          <PostForm onSubmit={this.onSubmit} /> 
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddPost: post => {
    dispatch(startAddPost(post));
  }
});

export default connect(
  undefined,
  mapDispatchToProps
)(AddPostPage);
