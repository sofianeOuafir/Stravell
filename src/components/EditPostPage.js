import React from "react";
import PostForm from "./PostForm";
import { connect } from "react-redux";
import Router, { withRouter } from "next/router";
import { slugify } from "underscore.string";

import PageHeader from "./PageHeader";
import {
  EDIT_POST_PAGE_DESCRIPTION,
  EDIT_POST_PAGE_TITLE
} from "./../constants/constants";
import Layout from "./Layout";
import { getPost } from "../queries/post";
import BreadCrumb from "./Breadcrumb";
import { addPost, removePost } from "./../queries/post";

export class EditPostPage extends React.Component {
  constructor(props) {
    super(props);
  }

  onSubmit = ({ post, country, user, place, region }) => {
    const { post: postBeforeUpdate, uid, userName } = this.props;

    removePost(postBeforeUpdate)
      .then(() => {
        return addPost({ post, country, user, place, region });
      })
      .then(() => {
        this.props.router.push(
          `/dashboard?uid=${uid}`,
          `/dashboard/${slugify(userName)}/${uid}`
        );
      });
  };

  render() {
    const { post, userName, uid } = this.props;
    const breadcrumbLinks = [
      { href: "/", text: "Home" },
      {
        href: `/dashboard/${slugify(userName)}/${uid}`,
        text: "Dashboard"
      },
      {
        href: `/editPost?id=${post.id}`,
        as: `/p/edit/${slugify(post.title)}/${post.id}`,
        text: "Edit Post",
        active: true
      }
    ];
    return (
      <Layout
        title={EDIT_POST_PAGE_TITLE}
        description={EDIT_POST_PAGE_DESCRIPTION}
      >
        <PageHeader title="Edit Post" withSocialShareButtons={false} />
        <div className="content-container">
          <div className="mb1">
            <BreadCrumb links={breadcrumbLinks} />
          </div>

          <PostForm post={post} onSubmit={this.onSubmit} />
        </div>
      </Layout>
    );
  }
}

EditPostPage.getInitialProps = async function({ query, req, reduxStore, res }) {
  const { id } = query;
  const post = await getPost(id);

  let authorised = false;
  if (req && req.session.decodedToken) {
    const user = req.session.decodedToken;
    if (user.user_id === post.uid) {
      authorised = true;
    }
  } else {
    if (reduxStore.getState().auth.uid === post.uid) {
      authorised = true;
    }
  }

  if (authorised) {
    return { post };
  } else {
    if (res) {
      res.writeHead(302, {
        Location: "/"
      });
      res.end();
    } else {
      Router.push("/");
    }
  }
};

const mapStateToProps = state => ({
  uid: state.auth.uid,
  userName: state.auth.userName
});

export default connect(mapStateToProps)(withRouter(EditPostPage));
