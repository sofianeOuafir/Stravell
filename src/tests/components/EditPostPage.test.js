import React from "react";
import { shallow } from "enzyme";
import { EditPostPage } from "./../../components/EditPostPage";
import PostForm from "./../../components/PostForm";
import posts from "./../fixtures/posts";

const removePost = jest.fn(() => {
  return Promise.resolve();
});
const addPost = jest.fn(() => {
  return Promise.resolve();
});
const router = { push: jest.fn() };
const post = posts[0];
const uid = 1;
const userName = "Sofiane";
const props = { removePost, addPost, router, post, uid, userName };
let wrapper;

beforeEach(() => {
  wrapper = shallow(<EditPostPage {...props} />);
});

test("should render correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should edit the post and redirect the user to the dashboard when the form get submitted", done => {
  wrapper.find(PostForm).simulate("submit", {});
  new Promise((resolve, reject) => {
    expect(removePost).toHaveBeenCalled();
    resolve();
  }).then(() => {
    expect(addPost).toHaveBeenCalled();
  })
});
