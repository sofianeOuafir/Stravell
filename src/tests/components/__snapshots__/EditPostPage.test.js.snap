// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<Layout
  description="Stravell | This page allow you to edit a travel post, article you've written."
  title="Stravell | Edit Post"
>
  <PageHeader
    title="Edit Post"
  />
  <div
    className="content-container"
  >
    <PostForm
      onSubmit={[Function]}
      post={
        Object {
          "id": 1,
        }
      }
    />
  </div>
</Layout>
`;
